test/test_rpc.py

```diff
@@ -648,3 +654,364 @@ def never_stops():
+            assert echo(3) == 3
+
+    @pytest.mark.enable_retry
+    def test_with_retry_policy(
+        self, client_container, toxiproxy, use_confirms
+    ):
+        if not use_confirms:
+            pytest.skip(
+                "unconfirmed messages will be lost after disconnection"
+            )
+
+        with entrypoint_hook(client_container, 'echo') as echo:
+            assert echo(1) == 1
+
+        toxiproxy.disable()
+
+        def enable_after_retry(args, kwargs, res, exc_info):
+            toxiproxy.enable()
+            return True
+
+        # call 2 succeeds (after reconnecting via retry policy)
+        with patch_wait(Connection, 'connect', callback=enable_after_retry):
+            with entrypoint_hook(client_container, 'echo') as echo:
+                assert echo(2) == 2
+
+
+@skip_if_no_toxiproxy
+class TestResponderDisconnections(object):
+    """ Test and demonstrate behaviour under poor network conditions.
+
+    Publisher confirms must be enabled for some of these tests to pass. Without
+    confirms, previously used but now dead connections will accept writes
+    without raising. These tests are skipped in this scenario.
+
+    Note that publisher confirms do not protect against sockets that remain
+    open but do not deliver messages (i.e. `toxiproxy.timeout(0)`).
+    This can only be mitigated with AMQP heartbeats (not yet supported)
+    """
+
+    @pytest.yield_fixture(autouse=True)
+    def toxic_responder(self, toxiproxy):
+        with patch.object(Responder, 'amqp_uri', new=toxiproxy.uri):
+            yield
+
+    @pytest.yield_fixture(autouse=True, params=[True, False])
+    def use_confirms(self, request):
+        with patch.object(Responder, 'use_confirms', new=request.param):
+            yield request.param
+
+    @pytest.yield_fixture(autouse=True)
+    def retry(self, request):
+        retry = False
+        if "enable_retry" in request.keywords:
+            retry = True
+
+        with patch.object(Responder, 'retry', new=retry):
+            yield
+
+    @pytest.fixture
+    def service_cls(self):
+        class Service(object):
+            name = "service"
+
+            @rpc
+            def echo(self, arg):
+                return arg
+
+        return Service
+
+    @pytest.fixture(autouse=True)
+    def container(self, container_factory, service_cls, rabbit_config):
+
+        config = rabbit_config
+
+        container = container_factory(service_cls, config)
+        container.start()
+        return container
+
+    @pytest.yield_fixture
+    def service_rpc(self, rabbit_config):
+
+        gts = []
+
+        def kill_greenthreads():
+            for gt in gts:
+                try:
+                    gt.kill()
+                except GreenletExit:  # pragma: no cover
+                    pass
+            del gts[:]
+
+        class ThreadSafeRpcProxy(object):
+            """ The ServiceRpcProxy is not thread-safe, so we can't create it
```

Comment:
Eventlet throws `Second simultaneous read` errors. I don't think we need to fix it.

The `ServiceRpcProxy` is synchronous and blocking so it won't interfere with whatever concurrency framework a client app uses. It assumes that it's being used either in synchronous code or safe concurrent code, whereas here we're sharing a single instance between multiple threads.